interface/wx/regex.h

```diff
@@ -250,5 +250,35 @@ class wxRegEx
         Replace the first occurrence.
     */
     int ReplaceFirst(wxString* text, const wxString& replacement) const;
+
+    /**
+        Escapes any of the following \, ^, $, ., |, ?, *, +, (, ), [, ], { and } metacharacters
+        found in the passed string with a backslash.
+
+        @param str
+            A string that may contain metacharacters to escape.
+
+        @return A string with all metacharacters converted to their escaped form.
+
+        @note This is a static function, and there is no need to create an instance of wxRegEx to use it.
+
+        @since 3.1.3
+    */
+    static wxString Escape(const wxString& str);
+
+    /**
+        Unescapes, that is removing any preceding backslash in any of the following
+        \, ^, $, ., |, ?, *, +, (, ), [, ], { and } metacharacters found in the passed string.
+
+        @param str
+            A string that may contain escaped metacharacters to unescape.
+
+        @return A string with any escaped metacharacter converted to its unescaped form.
```

Comment:
What happens if the original string wasn't properly escaped? It doesn't matter that much what do we do in this case, but we should at least define it.

Reply:
`wxRegEx::Unescape()` will unescape any escaped metacharacter of the passed string no matter what.

Reply:
Sorry for being unclear, by putting this question here I meant that the documentation should explain this.

To be honest, I'm not really sure how much sense does it make for `Unescape("\a")` to return the unchanged string (instead of `a`), but if this is how you want it to behave, I can live with it -- just please explain it in the docs.

Reply:
Please read this documentation [Unescape()](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.unescape?view=netframework-4.8) to understand.

I was thinking of a way to rephrase the .NET documentation, because it would be awkward to just pick their documentation and do it here.

As for the test I am working on it as you said take example from `tests/regex/wxregextest.cpp`

Reply:
.NET documentation says that the method throws an exception on invalid input, which is indeed the possibility I'd have chosen (in our case we don't use exceptions, but it could return `false`). But your code doesn't work like this...

Reply:
This version of `Unescape()` does not do the second part on the .NET one (the part talking about hexadecimal values), it is just exactly the reverse of `Escape()` and never fails or throws. And therefore, calling `Unescape(Escape(Unescape(Escape(Unescape(Escape(str))))));` will return exactly the initial string.

Reply:
Yes, this much is clear enough: `Unescape(Escape(s)) == s`. The question is, and was, since the beginning, about what happens if you pass it a string not created by `Escape()`, such as `\a` (`Escape()` would have turned it into `\\a`). There are several possibilities of what the function could do in this case:

1. Return `\a` unchanged.
2. Return just `a`.
3. Return an error.

and the documentation should really explain what happens (and the tests should check that it indeed happens).

Reply:
Aah,
You forgot that `Escape()` and `Unescape()` alters **only** the metacharacters that whould **cause a regex compilation error**, that is any one of these character in the static string: `static const wxString s_strMetaChars = wxT("\\^$.|?*+()[]{}");`, these are exacly 14 metacharacters: \\, ^, $, ., |, ?, *, +, (, ), [, ], { and }. And I have mentioned all these 14 metacharacters in the interface file. These are exacly the metacharacters used by **tcl** the one used by wxWidgets [here](https://www.tcl.tk/doc/howto/regexp81.tml), which differes slitly from other regex libraries like the .NET wich adds two more metacharacters (**#**, and **white space**).
any other character is not a metacharacter, thus `a` is not one, and wont be altered at all.

Given your example: `\a` when passed to `Escape()`, as you can see the string sample contains only one metacharacter which is the backslash, so it will be escaped by a backslash of course, then the for loop gets the next character which is `a` and is not a metacharacter, checked by the if statement:
```
        if (s_strMetaChars.find(*it) != wxString::npos)
        {
            strEscaped += wxT('\\');
        }
```
and thus will be left as is, (**unchanged**), and the final escaped string will be: `\\a`

The real purpose of this is when the developer expects string leterals from the end-user, where the end-user has nothing to do about regular expressions at all. He will be given an edit box or  `wxTextEntryDialog` for example. And to make this more clear lets pick the **Search and Replace** example and suppose wa have developed an application similar to Notepad++, then if the end-user wants to do a **Search and Replace** for example search for : this string: `:+? Hello ^$.` and replace it with: `wx regex works!`, you know that the end-user is free to enter what ever he wants, then in this case comes the power of `Escape()`, and the developer must call it passing the end-user string literals and then insert in the final **Search and Replace** regex to avoid compilation errors.

Reply:
We keep speaking past each other :-( I understand perfectly well the behaviour of `Escape()` and I also understand, as I previously wrote, the behaviour of `Unescape()` when it's passed a properly escaped string, such as previously returned by `Escape()`. Moreover, I also understand why do you want to have `Escape()`, although not necessarily why do you need `Unescape()`.

So perhaps the simplest would be to just drop the latter and rename `Escape()` to `QuoteMeta()` because this is what it does and it's a perfectly good function like this.

If you really want to keep `Unescape()`, however, you *must* define what does it for *all* strings, not just the properly escaped ones because it can be called with them. And then you need to choose one of the 3 approaches above and document and test the one you chose.

Please let me know if anything is still unclear. If in doubt, just remove `Unescape()`.

Reply:
Hello,
and sorry for the misunderstanding, and the delay,
You are completely right, `Escape()` should be renamed to `QuoteMeta()` because that's really what is means, and Unescape() should simply be dropped.
Because I think, on only **Qt** ant **.NET** have it named Escape(), whearas all the other toolkits or languages have the name **Quote()** or **QuoteMeta()** like **PERL** and **Golang** for example. As for `Unescape()`, I think, only **.NET** that have it, and this does not necessarily mean **wxWidgets** should have it too.